vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

```diff
@@ -0,0 +1,969 @@
+  linear_layer$weights,
+  list(linear_layer$w, linear_layer$b)
+))
+```
+
+Note you also have access to a quicker shortcut for adding weight to a layer:
+the `add_weight()` method:
+
+
+```{r}
+Linear(keras$layers$Layer) %py_class% {
+  initialize <- function(units = 32, input_dim = 32) {
+    super$initialize()
+    w_init <- tf$random_normal_initializer()
+    self$w <- self$add_weight(
+      shape = shape(input_dim, units),
+      initializer = "random_normal",
+      trainable = TRUE
+    )
+    self$b <- self$add_weight(
+      shape = shape(units),
+      initializer = "zeros",
+      trainable = TRUE
+    )
+  }
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+
+x <- tf$ones(shape(2, 2))
+linear_layer <- Linear(4, 2)
+y <- linear_layer(x)
+print(y)
+```
+
+## Layers can have non-trainable weights
+
+Besides trainable weights, you can add non-trainable weights to a layer as
+well. Such weights are meant not to be taken into account during
+backpropagation, when you are training the layer.
+
+Here's how to add and use a non-trainable weight:
+
+```{r}
+ComputeSum(keras$layers$Layer) %py_class% {
+  initialize <- function(input_dim) {
+    super$initialize()
+    self$total <- tf$Variable(
+      initial_value = tf$zeros(shape(input_dim)),
+      trainable = FALSE
+    )
+  }
+
+  call <- function(inputs) {
+    self$total$assign_add(tf$reduce_sum(inputs, axis = 0L))
+    self$total
+  }
+}
+
+x <- tf$ones(shape(2, 2))
+my_sum <- ComputeSum(2)
+y <- my_sum(x)
+print(as.numeric(y))
+y <- my_sum(x)
+print(as.numeric(y))
+```
+
+It's part of `layer$weights`, but it gets categorized as a non-trainable weight:
+
+```{r}
+cat("weights:", length(my_sum$weights), "\n")
+cat("non-trainable weights:", length(my_sum$non_trainable_weights), "\n")
+
+# It's not included in the trainable weights:
+cat("trainable_weights:", my_sum$trainable_weights, "\n")
+```
+
+## Best practice: deferring weight creation until the shape of the inputs is known
+
+Our `Linear` layer above took an `input_dim `argument that was used to compute
+the shape of the weights `w` and `b` in `initialize()`:
+
+
+```{r}
+Linear(keras$layers$Layer) %py_class% {
+  initialize <- function(units = 32, input_dim = 32) {
+    super$initialize()
+    self$w <- self$add_weight(
+      shape = shape(input_dim, units),
+      initializer = "random_normal",
+      trainable = TRUE
+    )
+    self$b <- self$add_weight(
+      shape = shape(units),
+      initializer = "zeros",
+      trainable = TRUE
+    )
+  }
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+```
+
+In many cases, you may not know in advance the size of your inputs, and you
+would like to lazily create weights when that value becomes known, some time
+after instantiating the layer.
+
+In the Keras API, we recommend creating layer weights in the `build(self,
+inputs_shape)` method of your layer. Like this:
+
+```{r}
+# Same as below, but using R6.
```

Comment:
Will you show both in the end? In this case, I'd start with the Python-ey way, because it's what you mostly use, and then say sth like: "If you use keras in a package, ..." etc